src/compute/cast/primitive_to.rs

```diff
@@ -17,15 +16,14 @@ use super::CastOptions;
 pub fn primitive_to_binary<T: NativeType + lexical_core::ToLexical, O: Offset>(
     from: &PrimitiveArray<T>,
 ) -> BinaryArray<O> {
-    let mut buffer = vec![];
     let builder = from.iter().fold(
         MutableBinaryArray::<O>::with_capacity(from.len()),
         |mut builder, x| {
             match x {
-                Some(x) => {
-                    lexical_to_bytes_mut(*x, &mut buffer);
-                    builder.push(Some(buffer.as_slice()));
-                }
+                Some(x) => unsafe {
+                    builder.reserve(1, T::FORMATTED_SIZE_DECIMAL);
+                    builder.write_values(|bytes| lexical_core::write(*x, bytes).len());
+                },
                 None => builder.push_null(),
```

Comment:
this is a really cool idea!

I think we may go a step further, though: since the size is constant, the offsets will be `[0, N, 2N, ..., M*N]` and the values can be constructed directly from `lexical_core::write`, e.g. via `extend`. We also do not need to check for utf8 below because lexical_core guarantees this. We can even ignore the validity of the primitive array and continue writing whatever is in the null slot, and clone the validity.

I think this implementation is best done without a `MutableBinaryArray`, though: we benefit from operating on the buffers directly in this case.

Reply:
>  ` since the size is constant, the offsets will be [0, N, 2N, ..., M*N]`

It's not constant,  `T::FORMATTED_SIZE_DECIMAL` is the maximum size to reverse.